c2rust-transpile/src/translator/assembly.rs

```diff
@@ -472,10 +472,55 @@ fn asm_is_att_syntax(asm: &str) -> bool {
     }
 }
 
+/// If applicable, maps the index of an input operand to the
+/// output operand it is tied to.
+///
+/// Utilizes the fact that GNU inline assembly specifies all
+/// operands in a particular order of [output1, ... outputN]
+/// followed by [input1, ..., inputN] where the indices for all
+/// input operands are indexed relative to the size of the
+/// output operand sequence.
+fn remapped_idx(
+    idx_str: &str,
+    num_output_operands: usize,
+    tied_operands: &HashMap<(usize, bool), usize>,
+) -> String {
+    let idx = idx_str
+        .trim_matches(|c: char| !c.is_ascii_digit())
+        .parse::<usize>()
+        .expect(format!("Could not parse '{:?}' as usize", idx_str).as_str()); // TODO: avoid panicking
+    if let Some(adj_idx) = idx.checked_sub(num_output_operands) {
+        // will only be Some(idx) if it was an input operand
+        match tied_operands.get(&(adj_idx, false)) {
+            Some(out_idx) => {
+                return out_idx.to_string();
+            }
+            None => {
+                // get number of tied inputs before this index
+                // TODO: can calculate this once before the call, not once for each input
+                let num_tied_before = tied_operands
+                    .keys()
+                    .filter(|&&(iidx, is_out)| !is_out && iidx < adj_idx)
+                    .count();
+                // shift the original index by the number of tied input operands prior to the current one
+                return (idx - num_tied_before).to_string();
+            }
+        };
+    }
+
+    idx_str.to_string()
+}
+
 /// Rewrite a LLVM inline assembly template string into an asm!-compatible one
 /// by translating its references to operands (of the form $0 or $x0) to {0} or
 /// {0:y} (and wrapping mem-only references in square brackets).
-fn rewrite_asm<F: Fn(&str) -> bool>(asm: &str, is_mem_only: F, arch: Arch) -> String {
+fn rewrite_asm<F: Fn(&str) -> bool>(
+    asm: &str,
+    num_output_operands: usize,
+    tied_operands: &HashMap<(usize, bool), usize>,
+    is_mem_only: F,
+    arch: Arch,
+) -> String {
```

Comment:
Given that `num_output_operands` and `tied_operands` are only used to remap given operand reference indices, I think it would be cleaner to accept an `Fn` closure that remaps operand indices here.